Project export

ActiveChooN opened the pull request:
<!---
Copyright (C) 2020-2021 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It will help avoiding duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

PR provides the ability to export the project as a dataset or annotation, reworked export menus.

Resolve #2911 
Resolve #2678 
Related #1278


![изображение](https://user-images.githubusercontent.com/8168727/125066586-d6ab7880-e0bb-11eb-9af7-de111ec09b61.png)

TODOs:
- [x] Fix image exporting in some cases
- [x] Add support for CVAT formats
- [x] Add server unit tests
- [x] Add UI support for exporting project and rework export task dataset menus

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [x] I have added a description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
~~- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly~~
- [x] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

zhiltsov-max commented on .vscode/launch.json:
```diff
@@ -7,7 +7,7 @@
         {
```
Probably, this file is not related to the PR

ActiveChooN replied:
Sure, will move the config improvement to a separate PR

zhiltsov-max commented on .vscode/settings.json:
```diff
@@ -1,5 +1,4 @@
 {
```
Probably, this file is not related to the PR

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -482,7 +503,339 @@ def match_frame_fuzzy(self, path):
+
+
+class CVATDataExtractor(datumaro.SourceExtractor):
+    dm_items: List[datumaro.DatasetItem] = []
```
This is antipattern - such declaration will share the value assigned here between instances. To fix, use the attrs library (https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/components/extractor.py#L31) or assign values in the `__init__` method.

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -482,7 +503,339 @@ def match_frame_fuzzy(self, path):
+    def __init__(self):
+        super().__init__()
+
+    def __iter__(self):
+        for item in self._items:
+            yield item
+
+    def __len__(self):
+        return len(self._items)
+
+    def categories(self):
+        return self._categories
```
These methods are defined in the base class.

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -482,7 +503,339 @@ def match_frame_fuzzy(self, path):
+
+    @staticmethod
+    def _load_categories(labels: list):
+        categories: Dict[datumaro.AnnotationType, datumaro.LabelCategories] = {}
```
```suggestion
        categories: Dict[datumaro.AnnotationType, datumaro.Categories] = {}
```

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -595,101 +939,189 @@ def _load_categories(cvat_anno, dimension):
+    pass
+
+def mung_image_name(name: str, images: List[Literal["path", "width", "height", "subset"]]) -> str:
+    name, ext = name.split(osp.extsep)
```
```suggestion
    name, ext = name.rsplit(osp.extsep, maxsplit=1)
```
It can produce more (and, theoretically, less) than 2 values: `/p.a.t.h/img.jpg.png.qq.zip.bmp`

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -595,101 +939,189 @@ def _load_categories(cvat_anno, dimension):
+def mung_image_name(name: str, images: List[Literal["path", "width", "height", "subset"]]) -> str:
+    name, ext = name.split(osp.extsep)
+    if not any(map(lambda image: image["path"] == name, images)):
+        return name + osp.extsep + ext
```
Maybe you'd want to use
```suggestion
        return osp.extsep.join(parts)
```

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -595,101 +939,189 @@ def _load_categories(cvat_anno, dimension):
+
+def mung_image_name(name: str, images: List[Literal["path", "width", "height", "subset"]]) -> str:
+    name, ext = name.split(osp.extsep)
+    if not any(map(lambda image: image["path"] == name, images)):
```
```suggestion
    if all(map(lambda image: image["path"] != name, images)):
```

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -595,101 +939,189 @@ def _load_categories(cvat_anno, dimension):
+    else:
+        i = 1
+        while i < sys.maxsize:
+            if not any(map(lambda image: image["path"] == f"{name}_{i}{osp.extsep}{ext}", images)):
```
Probably, it's a waste of CPU. This line has a complexity of O(task_size), and this is called in a loop. Maybe you can pass the image name dict and check if it contains a name?

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -595,101 +939,189 @@ def _load_categories(cvat_anno, dimension):
+                raise Exception(
+                    "Failed to convert attribute '%s'='%s': %s" %
+                    (a_name, a_value, e))
+            if format_name == "sly_pointcloud" and (a_desc.get('input_type') == 'select' or a_desc.get('input_type') == 'radio'):
```
This is not used and needs to be removed. Please, remove.

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -595,101 +939,189 @@ def _load_categories(cvat_anno, dimension):
+            if dimension == DimensionType.DIM_3D:
+                if format_name == "sly_pointcloud":
+                    anno_id = shapes[index]["id"]
+                    anno_attr["label_id"] = shapes[index]["label_id"]
```
This is not used.

ActiveChooN replied:
It came from Milestone 6 PR merge

zhiltsov-max replied:
I know. But it is extra.

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -595,101 +939,189 @@ def _load_categories(cvat_anno, dimension):
+
 
-            return dm_attr
+def GetCVATDataExtractor(instance_data: Union[ProjectData, TaskData], include_images: bool=False):
```
Probably, we need to find more descriptive term instead of "instance".

ActiveChooN replied:
Unfortunately, I can't think of a new one. I guess it's not a big problem, because it has a descriptive type. But if you have any ideas, I will change that.

zhiltsov-max commented on cvat/apps/dataset_manager/formats/velodynepoint.py:
```diff
@@ -20,6 +20,9 @@
 def _export_images(dst_file, task_data, save_images=False):
 
+    if not isinstance(task_data, TaskData):
+        raise Exception("Export to kitti raw format is working only with tasks")
```
Why?

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -579,9 +579,7 @@ def _init_frame_info(self):
-                    "path": mung_image_name(db_image.path,
-                        filter(lambda frame: frame["subset"] == task.subset ,self._frame_info.values())
-                    ),
+                    "path": mangle_image_name(db_image.path, self._frame_info.values(), task.subset),
```
I see this dict is built in O(task_size ^ 2) time. Maybe, we can find a better approach?

ActiveChooN commented:
@bsekachev, can you please also review the UI part of this PR?

bsekachev commented on cvat-ui/src/reducers/export-reducer.ts:
```diff
@@ -0,0 +1,103 @@
+            return {
+                ...state,
+                tasks: {
+                    datasets: instance instanceof core.classes.Task && saveImages ? activities : state.tasks.datasets,
+                    annotations:
+                        instance instanceof core.classes.Task && !saveImages ? activities : state.tasks.annotations,
+                },
+                projects: {
+                    datasets:
+                        instance instanceof core.classes.Project && saveImages ? activities : state.projects.datasets,
+                    annotations:
+                        instance instanceof core.classes.Project && !saveImages ?
+                            activities :
+                            state.projects.annotations,
```
You didn't make a deep copy for "activities", and added property to the object, so it is already in redux storage. And on the one hand, we could just destructure all the properties like this:
tasls: {
    datasets: ...tasks.datasets,
    annotations: ...tasks.annotations,
}

But I am not sure this way is also right, because actually we should not change original storage as it done in line 47.

ActiveChooN replied:
Added deep copy for activities state

bsekachev commented on cvat-ui/src/reducers/export-reducer.ts:
```diff
@@ -0,0 +1,103 @@
+            };
+        }
+        case ExportActionTypes.EXPORT_DATASET_FAILED:
+        case ExportActionTypes.EXPORT_DATASET_SUCCESS: {
```
Generally, I have the same thoughts like for ExportActionTypes.EXPORT_DATASET

bsekachev commented:
"Export Task dataset" does not work on annotation page.
I am not sure our DA team and community ok with the fact that we removed ability to dump annotations for a job (but we still have an ability to upload them). 

When I select export format in the modal, and then check "Save images", export format is reset

If I do not specify extension in a custom name, it will dump without it. I am not sure it is a good solution, especially for Windows users who are less experienced by default and where this file cannot be opened automatically. 

When we submit export, no any progress visible (it just closes modal window). I believe we should show any "status of waiting" somewhere for a user. In other case for huge tasks (or overloaded server) user thinks that someting is going wrong.

bsekachev commented:
One feature was lost.
When we dump/export annotations, default format (CVAT for videos, CVAT for images) is highliged depending of task type. Probably it is not applicable for datasets, but we could do it at least for tasks.

bsekachev commented on cvat-ui/src/components/export-dataset/export-dataset-modal.tsx:
```diff
@@ -0,0 +1,142 @@
+    };
+
+    return (
+        <Form
+            name='Export dataset'
+            form={form}
+            labelCol={{ span: 8 }}
+            wrapperCol={{ span: 16 }}
+            onValuesChange={handleValuesChange}
+            initialValues={
+                {
+                    selectedFormat: undefined,
+                    saveImages: false,
+                    customName: undefined,
+                } as FormValues
+            }
+            onFinish={handleExport}
+        >
+            <Modal
+                title={`Export ${instanceType} #${instance?.id} as a dataset`}
+                visible={modalVisible}
+                onCancel={closeModal}
+                onOk={() => form.submit()}
+            >
```
Looks unnatural, maybe better include Form to Modal, rather then Modal to Form?

bsekachev commented on cvat-ui/src/components/export-dataset/export-dataset-modal.tsx:
```diff
@@ -0,0 +1,142 @@
+                    <Checkbox>Save images</Checkbox>
+                </Form.Item>
+                <Form.Item label='Custom name' name='customName'>
+                    <Input placeholder='Custom name for dataset' />
```
```suggestion
                    <Input placeholder='Custom name for a dataset' />
```

bsekachev commented on cvat-ui/src/components/export-dataset/export-dataset-modal.tsx:
```diff
@@ -0,0 +1,142 @@
+                            .sort((a: any, b: any) => a.name.localeCompare(b.name))
+                            .filter(
+                                (dumper: any): boolean =>
+                                    !(instance instanceof core.classes.Task && dumper.dimension !== instance.dimension),
```
Difficult to understand.
Let's rewrite it as 

!(instance instanceof core.classes.Task) || dumper.dimension === instance.dimension

bsekachev commented on cvat-ui/src/components/export-dataset/export-dataset-modal.tsx:
```diff
@@ -0,0 +1,142 @@
+        initActivities();
+    }, [instance?.id, instance instanceof core.classes.Project]);
+
+    const closeModal = (): void => {
+        form.resetFields();
+        dispatch(exportActions.closeExportModal());
+    };
+
+    const handleValuesChange = (changedValues: any): void => {
+        if ('saveImages' in changedValues) {
+            form.setFieldsValue({ selectedFormat: undefined });
+        }
+        initActivities();
+    };
+
+    const handleExport = (values: FormValues): void => {
+        // have to validate format before so it would not be undefined
+        dispatch(
+            exportDatasetAsync(instance, values.selectedFormat as string, values.customName || '', values.saveImages),
+        );
+        closeModal();
+    };
```
Consider using useCallback hook

ActiveChooN commented on cvat-ui/src/reducers/export-reducer.ts:
```diff
@@ -0,0 +1,103 @@
+            const { instance, format, saveImages } = action.payload;
+            let activities;
+            if (instance instanceof core.classes.Project) {
+                activities = saveImages ? state.projects.datasets : state.projects.annotations;
```
It was in different groups in the original implementation so initially, I splat them too. But I agree with you.

ActiveChooN commented:
> "Export Task dataset" does not work on annotation page.

Fixed

> I am not sure our DA team and community ok with the fact that we removed ability to dump annotations for a job (but we still have an ability to upload them).

If I noted correctly dumping annotation for the job is just calling dumping annotation for the corresponding task, so that ability is present. But correct me if I'm wrong.

> When we submit export, no any progress visible (it just closes modal window). I believe we should show any "status of waiting" somewhere for a user.

It was done like in original implementation with loading circle in format select if dumping process is active.

bsekachev commented:
@ActiveChooN 

> It was done like in original implementation with loading circle in format select if dumping process is active.

Agree, but it's not always clear for a user, because they just do not see it. It would be better to use for example message component from Antd. Just an idea how to improve.

bsekachev commented:
> When I select export format in the modal, and then check "Save images", export format is reset

Still reproducable. I saw in the code that it was intended, but from the point of view UX imagine yourself a user, who is going to export annotations. First the user selects a format, then decided if he needs images, or not, and he notices how format is reset if the flag was changed.

bsekachev commented on cvat-core/src/project-implementation.js:
```diff
@@ -0,0 +1,74 @@
+// Copyright (C) 2019-2021 Intel Corporation
```
Right year is probably 2021, because it has been just created, right?

ActiveChooN commented:
> Still reproducable.

Fixed

> but it's not always clear for a user

Added notification

> When we dump/export annotations, default format (CVAT for videos, CVAT for images) is highliged depending of task type.

If the instance is a task and we have sufficient mode, the format is selected automatically.

bsekachev commented:
@ActiveChooN 
Please, fix CI tests

ActiveChooN commented:
@dvkruchinin, can you please adjust tests for the new dumping process?

bsekachev commented:
@dvkruchinin 

It is has higher priority then OpenCV.js testing

dvkruchinin commented:
> can you please adjust tests for the new dumping process?
> It is has higher priority then OpenCV.js testing

Sure. I'll take care of it.

nmanovic commented:
@ActiveChooN , after I create a task for a project, the field "project" is reset. Thus to create a new task in the project, I have to correct the field value. Is it expected?

nmanovic commented:
When export a dataset, got the following error. My project had 3 jobs and two subsets (train and test).

![image](https://user-images.githubusercontent.com/40690625/127359824-882652fc-22c1-4c9c-b4cd-2ff6f254b159.png)

```
ERROR - 2021-07-28 16:21:51,224 - worker - Traceback (most recent call last):
  File "/opt/venv/lib/python3.8/site-packages/rq/worker.py", line 936, in perform_job
    rv = job.perform()
  File "/opt/venv/lib/python3.8/site-packages/rq/job.py", line 684, in perform
    self._result = self._execute()
  File "/opt/venv/lib/python3.8/site-packages/rq/job.py", line 690, in _execute
    return self.func(*self.args, **self.kwargs)
  File "/home/django/cvat/apps/dataset_manager/views.py", line 103, in export_project_as_dataset
    return export(dst_format, project_id=project_id, server_url=server_url, save_images=True)
  File "/home/django/cvat/apps/dataset_manager/views.py", line 72, in export
    export_fn(db_instance.id, temp_file, dst_format,
  File "/home/django/cvat/apps/dataset_manager/project.py", line 29, in export_project
    project.export(f, exporter, host=server_url, save_images=save_images)
  File "/home/django/cvat/apps/dataset_manager/project.py", line 68, in export
    exporter(dst_file, project_data, **options)
  File "/home/django/cvat/apps/dataset_manager/formats/registry.py", line 37, in __call__
    f_or_cls(*args, **kwargs)
  File "/home/django/cvat/apps/dataset_manager/formats/coco.py", line 22, in _export
    dataset.export(temp_dir, 'coco_instances', save_images=save_images,
  File "/opt/venv/lib/python3.8/site-packages/datumaro/util/__init__.py", line 203, in wrapped_func
    func(*args, **kwargs)
  File "/opt/venv/lib/python3.8/site-packages/datumaro/components/dataset.py", line 774, in export
    converter.convert(self, save_dir=save_dir, **kwargs)
  File "/opt/venv/lib/python3.8/site-packages/datumaro/components/converter.py", line 33, in convert
    return converter.apply()
  File "/opt/venv/lib/python3.8/site-packages/datumaro/plugins/coco_format/converter.py", line 652, in apply
    self._save_image(item, subdir=osp.join(self._images_dir,
  File "/opt/venv/lib/python3.8/site-packages/datumaro/components/converter.py", line 90, in _save_image
    item.image.save(path)
  File "/opt/venv/lib/python3.8/site-packages/datumaro/util/image.py", line 383, in save
    f.write(self.get_bytes())
  File "/opt/venv/lib/python3.8/site-packages/datumaro/util/image.py", line 355, in get_bytes
    return self._bytes_data()
  File "/opt/venv/lib/python3.8/site-packages/datumaro/util/image.py", line 227, in __call__
    image = self.loader(self.path)
  File "/home/django/cvat/apps/dataset_manager/bindings.py", line 966, in <lambda>
    loader = lambda _: frame_provider.get_frame(i,
  File "/home/django/cvat/apps/engine/frame_provider.py", line 172, in get_frame
    _, chunk_number, frame_offset = self._validate_frame_number(frame_number)
  File "/home/django/cvat/apps/engine/frame_provider.py", line 122, in _validate_frame_number
    raise Exception('Incorrect requested frame number: {}'.format(frame_number_))
Exception: Incorrect requested frame number: 23
Traceback (most recent call last):
  File "/opt/venv/lib/python3.8/site-packages/rq/worker.py", line 936, in perform_job
    rv = job.perform()
  File "/opt/venv/lib/python3.8/site-packages/rq/job.py", line 684, in perform
    self._result = self._execute()
  File "/opt/venv/lib/python3.8/site-packages/rq/job.py", line 690, in _execute
    return self.func(*self.args, **self.kwargs)
  File "/home/django/cvat/apps/dataset_manager/views.py", line 103, in export_project_as_dataset
    return export(dst_format, project_id=project_id, server_url=server_url, save_images=True)
  File "/home/django/cvat/apps/dataset_manager/views.py", line 72, in export
    export_fn(db_instance.id, temp_file, dst_format,
  File "/home/django/cvat/apps/dataset_manager/project.py", line 29, in export_project
    project.export(f, exporter, host=server_url, save_images=save_images)
  File "/home/django/cvat/apps/dataset_manager/project.py", line 68, in export
    exporter(dst_file, project_data, **options)
  File "/home/django/cvat/apps/dataset_manager/formats/registry.py", line 37, in __call__
    f_or_cls(*args, **kwargs)
  File "/home/django/cvat/apps/dataset_manager/formats/coco.py", line 22, in _export
    dataset.export(temp_dir, 'coco_instances', save_images=save_images,
  File "/opt/venv/lib/python3.8/site-packages/datumaro/util/__init__.py", line 203, in wrapped_func
    func(*args, **kwargs)
  File "/opt/venv/lib/python3.8/site-packages/datumaro/components/dataset.py", line 774, in export
    converter.convert(self, save_dir=save_dir, **kwargs)
  File "/opt/venv/lib/python3.8/site-packages/datumaro/components/converter.py", line 33, in convert
    return converter.apply()
  File "/opt/venv/lib/python3.8/site-packages/datumaro/plugins/coco_format/converter.py", line 652, in apply
    self._save_image(item, subdir=osp.join(self._images_dir,
  File "/opt/venv/lib/python3.8/site-packages/datumaro/components/converter.py", line 90, in _save_image
    item.image.save(path)
  File "/opt/venv/lib/python3.8/site-packages/datumaro/util/image.py", line 383, in save
    f.write(self.get_bytes())
  File "/opt/venv/lib/python3.8/site-packages/datumaro/util/image.py", line 355, in get_bytes
    return self._bytes_data()
  File "/opt/venv/lib/python3.8/site-packages/datumaro/util/image.py", line 227, in __call__
    image = self.loader(self.path)
  File "/home/django/cvat/apps/dataset_manager/bindings.py", line 966, in <lambda>
    loader = lambda _: frame_provider.get_frame(i,
  File "/home/django/cvat/apps/engine/frame_provider.py", line 172, in get_frame
    _, chunk_number, frame_offset = self._validate_frame_number(frame_number)
  File "/home/django/cvat/apps/engine/frame_provider.py", line 122, in _validate_frame_number
    raise Exception('Incorrect requested frame number: {}'.format(frame_number_))
Exception: Incorrect requested frame number: 23
```

nmanovic commented:
@ActiveChooN , @bsekachev , what do you think if we rename menu items?

- [ ] Rename "Export task dataset" -> "Export"
- [ ] Rename "Export task" -> "Backup"
- [ ] Redo "Upload annotations" -> "Import" (need to have a similar dialog with ability to upload a file and choose dataset format).
- [ ] Rename "Move to project -> "Move"
- [ ] Rename "Automatic annotation" -> "Run AI"

ActiveChooN commented:
> 
> 
> @ActiveChooN , @bsekachev , what do you think if we rename menu items?
> 
>     * [ ]  Rename "Export task dataset" -> "Export"
> 
>     * [ ]  Rename "Export task" -> "Backup"
> 
>     * [ ]  Redo "Upload annotations" -> "Import" (need to have a similar dialog with ability to upload a file and choose dataset format).
> 
>     * [ ]  Rename "Move to project -> "Move"
> 
>     * [ ]  Rename "Automatic annotation" -> "Run AI"

@nmanovic, I agree, but maybe better

- "Automatic annotation" -> "Run AI annotation"

bsekachev commented:
@nmanovic 

> Rename "Export task dataset" -> "Export"

Generally, agree. 

> Rename "Export task" -> "Backup"

Backup is a narrower term, than the feature covers. And if we change it to backup, we need a new term for "Import Task". More specific term might be "Export CVAT task"/"Import CVAT task"

> Redo "Upload annotations" -> "Import" (need to have a similar dialog with ability to upload a file and choose dataset format).

Generally, agree. 

> Rename "Move to project -> "Move"

"Move" does not provide enough details 

> Rename "Automatic annotation" -> "Run AI"

Might be.

dvkruchinin commented:
When exporting a task, the variable ```instanceType``` in the notification does not get a value immediately. Only when re-exporting.
![export_datataset_notification_issue](https://user-images.githubusercontent.com/33020454/127458369-4281846e-44d3-47ad-9320-f12874a0287e.gif)

ActiveChooN commented:
@dvkruchinin 

> When exporting a task, the variable `instanceType` in the notification does not get a value immediately. Only when re-exporting.

Fixed

@nmanovic 

> @ActiveChooN, after I create a task for a project, the field "project" is reset. Thus to create a new task in the project, I have to correct the field value. Is it expected?

Patch that fixed it was in develop branch, but not in my branch. Now should be ok.

dvkruchinin commented:
When exporting a task for a 3D functional, the default format value is ```CVAT for images 1.1```.
![export_task_3d_format_issue](https://user-images.githubusercontent.com/33020454/127493437-84f43f1f-f093-49f1-8f1d-de3dac687b40.JPG)
Probably this is not quite correct.

ActiveChooN commented:
> When export a dataset, got the following error. My project had 3 jobs and two subsets (train and test).

@nmanovic, fixed



> When exporting a task for a 3D functional, the default format value is `CVAT for images 1.1`.

@dvkruchinin, fixed

nmanovic commented on .github/workflows/main.yml:
```diff
@@ -5,8 +5,7 @@ on:
       - 'develop'
   pull_request:
-    branches:
-      - '*'
```
@ActiveChooN , what did you make the change?

ActiveChooN replied:
@dvkruchinin  is the original author. Dmitry, can you comment, please?

dvkruchinin replied:
Without these changes, tests are not run if a PR is created for another PR. I added this here to make it clear that the adapted tests are successful for this PR. These changes can also be made by a separate PR.

nmanovic commented on cvat/apps/dataset_manager/views.py:
```diff
@@ -29,62 +30,81 @@ def log_exception(logger=None, exc_info=True):
+    return export(dst_format, project_id=project_id, server_url=server_url, save_images=True)
+
+
+def export_project_annotation(project_id, dst_format=None, server_url=None):
```
```suggestion
def export_project_annotations(project_id, dst_format=None, server_url=None):
```

nmanovic commented:
@ActiveChooN , I have a project with two subsets: train with 2 tasks and test with 1 task. When I export as a COCO dataset, I have only one subset "default. Is it expected?

ActiveChooN commented:
> @ActiveChooN , we discussed that 3D support should be a separate PR. Now tests are failed again. Could you please revert changes?

Reverted

> @ActiveChooN , I have a project with two subsets: train with 2 tasks and test with 1 task. When I export as a COCO dataset, I have only one subset "default. Is it expected?

I guess it's not normal. I'm investigating it.

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -482,7 +503,327 @@ def match_frame_fuzzy(self, path):
+class CVATDataExtractor(datumaro.SourceExtractor):
+    def __init__(self):
+        super().__init__()
+        self.dm_items: List[datumaro.DatasetItem] = []
```
This variable is not used.

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -482,7 +503,327 @@ def match_frame_fuzzy(self, path):
+        return osp.splitext(path)[0]
+
+
+class CVATDataExtractor(datumaro.SourceExtractor):
```
The problem with subsets is because of the base class / project extractor does not update self._subsets.

ActiveChooN replied:
Thanks for describing

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -482,7 +503,327 @@ def match_frame_fuzzy(self, path):
+    def __init__(self):
+        super().__init__()
+        self.dm_items: List[datumaro.DatasetItem] = []
+        self._categories: Dict[datumaro.AnnotationType, datumaro.LabelCategories] = {}
```
```suggestion
        self._categories: Dict[datumaro.AnnotationType, datumaro.Categories] = {}
```

ActiveChooN commented:
> Hi, I've got an error after I created a project, moved a task with the same labels and then tried to move another task:

Thanks, @bsekachev already has reported the issue in #3403

ActiveChooN commented:
> Another problem: project dataset export cache reuses the cached version if I change the subset name of a task or any task annotations are changed.

@zhiltsov-max, fixed

gianrubio commented:
I just tested the new feature and it worked like a charm. Good job @ActiveChooN

ActiveChooN commented:
@aschernov, could you update documentation for the task dumping function and add documentation for the project export feature, please?

@dvkruchinin, could you also add tests for the project export function, please?

dvkruchinin commented:
> could you also add tests for the project export function, please?

Sure. A test will be prepared.

aschernov commented:
@ActiveChooN , sure. We'll update the documentation

jovialio commented:
> I just tested the new feature and it worked like a charm. Good job @ActiveChooN

@gianrubio How did you install the new feature. Seems like the feature is in v1.6 and yet to be released. Would love to have the Project exportation feature.

gianrubio commented:
You have to checkout the development branch and build your own docker
container. That all you have to do.

On Fri, 20 Aug 2021 at 09:05, Dennis Chew ***@***.***> wrote:

> I just tested the new feature and it worked like a charm. Good job
> @ActiveChooN <https://github.com/ActiveChooN>
>
> @gianrubio <https://github.com/gianrubio> How did you install the new
> feature. Seems like the feature is in v1.6 and yet to be released. Would
> love to have the Project exportation feature.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openvinotoolkit/cvat/pull/3365#issuecomment-902454443>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLOLDAUTVPOFFIAAIBZHBTT5XWCHANCNFSM47H5DP3A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>
-- 
Giancarlo Rubio

jovialio commented:
@gianrubio Got it! Thanks!

davodogster commented:
@aschernov Hi when will project export be avaliable in the non-development version of cvat? I don't know how to install the dev version and I don't want to install it. Cheers, Sam

bsekachev commented:
@davodogster 

Unfortunately, I can't say precisely, but the next release may come after 3-4 weeks.
I would emphasize, that the development branch is usually quite stable to real use, so if you need to use some features urgently, using dev might be an option.

davodogster commented:
@ActiveChooN Ok I'm keen to give the dev version a try. Now, When I do "git clone cvat" - does it clone the master branch?

1. Do I need to force it to clone the development branch instead of the master branch?
2. What is the docker-compose.dev.yml file for and do I need to **run that in addition to** docker-compose.override ??
Thanks! Sam

ActiveChooN commented:
Hi, @davodogster. 
- `git clone cvat` command clones repo with default develop branch.
- No, develop branch is the default for this repo.
- `docker-compose.dev.yml` is used for building a developing version of containers on your own and you should use it as an addition for `docker build` command.
- No, you don't lose anything, because data stored in docker volumes.

davodogster commented:
@ActiveChooN @zhiltsov-max Hi I just ran docker compose down cvat, then git cloned cvat using "git clone https://github.com/openvinotoolkit/cvat.git" and I have two problems.

1. Export project as dataset still does not appear as an option - The only project option is "delete"
2. It didn't solve this new error I have been getting https://github.com/openvinotoolkit/cvat.git (I tried the manually_migrate.py solution and it didn't work)
![image](https://user-images.githubusercontent.com/46079516/132440451-159cf0b0-e96e-4414-b367-f4147c551221.png)

I'm using a share folder with 549 large images. The task was successfully created and I can scroll across the images and annotate them, but the errors keep popping up.

Thank you, Sam
